File: ui/src/lib/api/teams.ts
import { request } from "./client";
import type { Membership, MemberRole, Project, Team } from "@/types/project";

/** `GET /teams` — every team (open read). */
export function listTeams(signal?: AbortSignal): Promise<Team[]> {
  return request<Team[]>("/teams", { signal });
}

/** `GET /teams/:id` — one team; `404` if absent. */
export function getTeam(id: string, signal?: AbortSignal): Promise<Team> {
  return request<Team>(`/teams/${encodeURIComponent(id)}`, { signal });
}

/** `POST /teams` — create (or re-affirm) a team. Requires `Author` + admin. */
export function createTeam(id: string, title: string): Promise<Team> {
  return request<Team>("/teams", { method: "POST", auth: true, body: { id, title } });
}

/** `GET /teams/:id/projects` — the projects placed `under` this team. */
export function listTeamProjects(id: string, signal?: AbortSignal): Promise<Project[]> {
  return request<Project[]>(`/teams/${encodeURIComponent(id)}/projects`, { signal });
}

/** `GET /teams/:id/members` — the team's members with their per-team roles. */
export function listMembers(team: string, signal?: AbortSignal): Promise<Membership[]> {
  return request<Membership[]>(`/teams/${encodeURIComponent(team)}/members`, { signal });
}

/** `POST /teams/:id/members` — add (or re-affirm) a member with a role. Requires
 *  `Author` + admin. */
export function addMember(team: string, user: string, role: MemberRole): Promise<Membership> {
  return request<Membership>(`/teams/${encodeURIComponent(team)}/members`, {
    method: "POST",
    auth: true,
    body: { user, role },
  });
}

/** `DELETE /teams/:id/members/:user` — remove a membership. Requires `Author` +
 *  admin. */
export function removeMember(team: string, user: string): Promise<{ removed: boolean }> {
  return request<{ removed: boolean }>(
    `/teams/${encodeURIComponent(team)}/members/${encodeURIComponent(user)}`,
    { method: "DELETE", auth: true },
  );
}

File: crates/lazybones-api/src/routes/teams.rs
//! `/teams` — the mid container of the team graph (a team owns projects, has
//! members).
//!
//! List/get/create plus the downward traversal that lists a team's projects
//! (`project ->under-> team`). Reads are open; creating a team is an
//! administrative act — it requires [`Capability::Author`] *and* clears the
//! org-graph role guard as an admin ([`RoleReq::Admin`]). In local
//! single-operator mode the role guard no-ops.

use axum::Json;
use axum::extract::{Path, State};
use lazybones_auth::Capability;
use lazybones_store::{Project, StoreError, Team};

use crate::dto::CreateTeamBody;
use crate::error::ApiResult;
use crate::extract::Session;
use super::guard::{RoleReq, ensure_role};
use crate::state::AppState;

/// `GET /teams` — list every team (open read).
pub async fn list_teams(State(state): State<AppState>) -> ApiResult<Json<Vec<Team>>> {
    Ok(Json(state.store.list_teams().await?))
}

/// `POST /teams` — create (or re-affirm) a team. Requires `Author` + admin.
/// Idempotent on the id (the org graph is cloud-authored, single-writer).
pub async fn create_team(
    State(state): State<AppState>,
    session: Session,
    Json(body): Json<CreateTeamBody>,
) -> ApiResult<Json<Team>> {
    session.require(Capability::Author, "author", &body.id)?;
    ensure_role(&state, &session, RoleReq::Admin).await?;

    let team = Team::new(&body.id, &body.title, state.store.now());
    Ok(Json(state.store.create_team(&team).await?))
}

/// `GET /teams/:id` — fetch one team (open read), or `404`.
pub async fn get_team(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> ApiResult<Json<Team>> {
    Ok(Json(
        state
            .store
            .get_team(&id)
            .await?
            .ok_or_else(|| StoreError::TeamNotFound(id.clone()))?,
    ))
}

/// `GET /teams/:id/projects` — the projects placed `under` this team, via the
/// containment traversal (open read). An unknown team simply lists empty.
pub async fn list_team_projects(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> ApiResult<Json<Vec<Project>>> {
    Ok(Json(state.store.team_projects(&id).await?))
}
